Zend/zend_compile.c

```diff
@@ -9443,6 +9443,35 @@ static void zend_compile_include_or_eval(znode *result, zend_ast *ast) /* {{{ */
 }
 /* }}} */
 
+static void zend_compile_nameof(znode *result, zend_ast *ast) /* {{{ */
+{
+	zend_ast *var_ast = ast->child[0];
+	zend_ast *name_ast;
+
+	// todo: proper error handling and better error messages
+
+	switch (var_ast->kind) {
+		case ZEND_AST_VAR:
+		case ZEND_AST_CALL:
+			name_ast = var_ast->child[0];
+			ZVAL_COPY(&result->u.constant, zend_ast_get_zval(name_ast));
+			result->op_type = IS_CONST;
+			break;
+		case ZEND_AST_PROP:
+		case ZEND_AST_NULLSAFE_PROP:
+		case ZEND_AST_STATIC_PROP:
+		case ZEND_AST_CLASS_CONST:
+			name_ast = var_ast->child[1];
+			ZVAL_COPY(&result->u.constant, zend_ast_get_zval(name_ast));
+			result->op_type = IS_CONST;
+			break;
+		default:
+			zend_error_noreturn(E_COMPILE_ERROR,
+				"Cannot use nameof() on the result of an expression or something with an ambiguous name");
```

Comment:
Can "something" be more specific in this error, or is it as specific as it can be?

Reply:
Yes and yes; that's actually a good question. I imagine that given more examples, more helpful error messages can be created, but there are "somethings" that are ambiguous or don't make sense:

```
echo nameof($a['b']);
echo nameof(empty($x));
echo nameof($x === $y);
echo nameof('literal');
```

In the grammar, I chose an expression as what can go in the brackets/parenthesis, but it doesn't make sense to accept literally any expression. Perhaps there is something better to use (or construct), but I didn't want to spend too much time on it in case there was a name for these more ambiguous things that I don't have the imagination for.